source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,702 @@
+      for (auto var : *replacements) {
+        std::unique_ptr<ir::Instruction> annotation(new ir::Instruction(
+            context(), SpvOpDecorate, 0, 0,
+            std::initializer_list<ir::Operand>{
+                {SPV_OPERAND_TYPE_ID, {var->result_id()}},
+                {SPV_OPERAND_TYPE_DECORATION, {decoration}}}));
+        for (uint32_t i = 2; i < inst->NumInOperands(); ++i) {
+          ir::Operand copy(inst->GetInOperand(i));
+          annotation->AddOperand(std::move(copy));
+        }
+        context()->AddAnnotationInst(std::move(annotation));
+        get_def_use_mgr()->AnalyzeInstUse(&*--context()->annotation_end());
+      }
+    }
+  }
+}
+
+void ScalarReplacementPass::CreateVariable(
+    uint32_t typeId, ir::Instruction* varInst, uint32_t index,
+    std::vector<ir::Instruction*>* replacements) {
+  uint32_t ptrId = GetOrCreatePointerType(typeId);
+  uint32_t id = TakeNextId();
+  std::unique_ptr<ir::Instruction> variable(new ir::Instruction(
+      context(), SpvOpVariable, ptrId, id,
+      std::initializer_list<ir::Operand>{
+          {SPV_OPERAND_TYPE_STORAGE_CLASS, {SpvStorageClassFunction}}}));
+
+  ir::BasicBlock* block = context()->get_instr_block(varInst);
+  block->begin().InsertBefore(std::move(variable));
+  ir::Instruction* inst = &*block->begin();
+
+  // If varInst was initialized, make sure to initialize its replacement.
+  GetOrCreateInitialValue(varInst, index, inst);
+  get_def_use_mgr()->AnalyzeInstDefUse(inst);
+  context()->set_instr_block(inst, block);
+
+  replacements->push_back(inst);
+}
+
+uint32_t ScalarReplacementPass::GetOrCreatePointerType(uint32_t id) {
+  auto iter = pointee_to_pointer_.find(id);
+  if (iter != pointee_to_pointer_.end()) return iter->second;
+
+  uint32_t ptrId = 0;
+  for (auto global : context()->types_values()) {
+    if (global.opcode() == SpvOpTypePointer &&
+        global.GetSingleWordInOperand(0u) == SpvStorageClassFunction &&
+        global.GetSingleWordInOperand(1u) == id) {
+      if (!context()->get_feature_mgr()->HasExtension(
+              libspirv::Extension::kSPV_KHR_variable_pointers) ||
+          get_decoration_mgr()->GetDecorationsFor(id, false).empty()) {
+        // If variable pointers is enabled, only reuse a decoration-less
+        // pointer of the correct type
+        ptrId = global.result_id();
+        break;
+      }
+    }
+  }
+
+  if (ptrId != 0) {
+    pointee_to_pointer_[id] = ptrId;
+    return ptrId;
+  }
+
+  ptrId = TakeNextId();
+  context()->AddType(MakeUnique<ir::Instruction>(
+      context(), SpvOpTypePointer, 0, ptrId,
+      std::initializer_list<ir::Operand>{
+          {SPV_OPERAND_TYPE_STORAGE_CLASS, {SpvStorageClassFunction}},
+          {SPV_OPERAND_TYPE_ID, {id}}}));
+  ir::Instruction* ptr = &*--context()->types_values_end();
+  get_def_use_mgr()->AnalyzeInstDefUse(ptr);
+  pointee_to_pointer_[id] = ptrId;
+
+  return ptrId;
+}
+
+uint32_t ScalarReplacementPass::GetIntId() {
```

Comment:
Likewise with this one.  The TypeManager would give you an existing Int, if it finds one (or should ;)

Reply:
Same as above, I'd prefer to leave this refactor to another PR.

Reply:
This dead, I've removed it.